File: Logistic Regression/logisticRegression.m
%% Machine Learning - Exercise 2: Logistic Regression
%
% INITIALIZE
clear;
close all;
clc;

data = load('data/ex2data1.txt');
X = data(:, [1, 2]);
y = data(:, 3);

addpath("octave")
% ====================   Part 1: Plotting   ====================
fprintf(['Plotting data with + indicating (y = 1) examples and o ' ...
         'indicating (y = 0) examples.\n']);

%FILE: plotData.m
plotData(X, y);
hold on;
xlabel('Exam 1 score')
ylabel('Exam 2 score')
legend('Admitted', 'Not admitted')
hold off;
fprintf('Paused. Press enter to continue.\n\n');
pause;


% ============== Part 2: Compute Cost and Gradient ==============
[m, n] = size(X);
X = [ones(m, 1), X];

%FILE: costFunction.m
% Compute and display initial cost and gradient
initial_theta = zeros(n + 1, 1);
[cost, grad] = costFunction(initial_theta, X, y);
fprintf('Cost at initial theta (zeros): %f\n', cost);
fprintf('Expected cost (approx): 0.693\n');
fprintf('Gradient at initial theta (zeros): \n');
fprintf(' %f \n', grad);
fprintf('Expected gradients (approx):\n -0.1000\n -12.0092\n -11.2628\n');

% Compute and display cost and gradient with non-zero theta
test_theta = [-24; 0.2; 0.2];
[cost, grad] = costFunction(test_theta, X, y);
fprintf('\nCost at test theta: %f\n', cost);
fprintf('Expected cost (approx): 0.218\n');
fprintf('Gradient at test theta: \n');
fprintf(' %f \n', grad);
fprintf('Expected gradients (approx):\n 0.043\n 2.566\n 2.647\n');

fprintf('\nProgram paused. Press enter to continue.\n');
pause;

%% ============= Optimizing using fminunc  =============
% Create options structure for optimization function
% In this case, this function is fminunc
options = optimset('GradObj', 'on', % Second arg returned will be the first derivative of gradient decent
                     'MaxIter', 400);

%  Optimization function fminun tries to determine the local minimum of a given function (in this case, a cost function)
[theta, cost] = fminunc( @(t)( costFunction(t, X, y) ),
                        initial_theta, options );

% Print theta to screen
fprintf('Cost at theta found by fminunc: %f\n', cost);
fprintf('Expected cost (approx): 0.203\n');
fprintf('theta: \n');
fprintf(' %f \n', theta);
fprintf('Expected theta (approx):\n');
fprintf(' -25.161\n 0.206\n 0.201\n');

%FILE: plotDecisionBoundary.m
plotDecisionBoundary(theta, X, y);

% Put some labels
hold on;
% Labels and Legend
xlabel('Exam 1 score')
ylabel('Exam 2 score')

% Specified in plot order
legend('Admitted', 'Not admitted')
hold off;

fprintf('\nProgram paused. Press enter to continue.\n');
pause;


%% ============== Part 4: Predict and Accuracies ==============
%  Predict probability for a student with score 45 on exam 1
%  and score 85 on exam 2
prob = sigmoid([1 45 85] * theta);
fprintf(['For a student with scores 45 and 85, we predict an admission ' ...
         'probability of %f\n'], prob);
fprintf('Expected value: 0.775 +/- 0.002\n\n');

% Compute accuracy on our training set
p = predict(theta, X);

fprintf('Train Accuracy: %f\n', mean(double(p == y)) * 100);
fprintf('Expected accuracy (approx): 89.0\n');
fprintf('\n');
fprintf('\nProgram paused. Press enter to close.\n');
pause

File: Logistic Regression/logisticRegression.py
## Logistic Regression
#
# INITIALIZE
import os
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
os.system('cls' if os.name == 'nt' else 'clear')
plt.close("all")

data = pd.read_csv('data/ex2data1.txt', names=['Exam 1 Score', 'Exam 2 Score', 'Admitted'])
X = data.iloc[:, 0:2]
y = data.iloc[:, 2]

# ================ Plotting ================
print(['Plotting data with + indicating (y = 1)',
        ' and o indicating (y = 0).\n'])

from python.plotData import plotData
plotData(X, y)

plt.ylabel('Exam 2 score')
plt.xlabel('Exam 1 score')

plt.show(block=False)

plt.legend(['Admitted', 'Not Admitted'])
input('Paused. Press enter to continue.\n')

# ============== Compute Cost and Gradient ==============
m, n = X.shape
X = np.append( np.ones((m, 1)), X.to_numpy().reshape(X.shape), axis=1)
y = y.to_numpy().reshape(y.size, 1)

from python.costFunction import costFunction
# Compute and display initial cost and gradient
initial_theta = np.zeros((n + 1, 1))
cost, grad = costFunction(X, y, initial_theta)
print('Cost at initial theta (zeros): \n', cost)
print('Expected cost (approx): 0.693')
print('Gradient at initial theta (zeros): \n', grad)
print('Expected gradients (approx):\n -0.1000\n -12.0092\n -11.2628\n')

# Compute and display cost and gradient with non-zero theta
test_theta = np.c_[[-24, 0.2, 0.2]]
cost, grad = costFunction(X, y, test_theta)
print('\nCost at test theta: ', cost)
print('Expected cost (approx): 0.218\n')
print('Gradient at test theta: \n', grad)
print('Expected gradients (approx):\n 0.043\n 2.566\n 2.647\n')

input('Paused. Press enter to continue.\n')

# ============= Optimizing using fminunc  =============
# Create options structure for optimization function
# In this case, this function is fminunc
options = { 
 "MaxIter": 400, 
 "GradObj": True # seek fminunc to also return the first derivative of gradient decent as a second arg returned will be
}

#  Optimization function fminun tries to determine the local minimum of a given function (in this case, a cost function)
from python.fminunc import fminunc
theta, cost = fminunc( lambda theta_: costFunction(X, y, theta_),
                        initial_theta, 
                        options )

# Print theta to screen
print('Cost at theta found by fminunc: \n', cost);
print('Expected cost (approx): 0.203');
print('theta: \n', theta);
print('Expected theta (approx):');
print(' -25.161\n 0.206\n 0.201\n');
# TODO